Fix issue with -fvirtual and using

leakec opened the pull request:
#2872 is for a parsing problem for a corner case where:
* A member is included on a grandchild class (call this `D2` for derived 2) via a `using` directive that points to the parent class (let's call this `D1` for derived 1). The parent does not do any specialization, so it's version of this method is the same as the `base` class (this is the grandparent class of `D2` and the parent class of `D1`).
* The `-fvirtual` flag is used. As a result, when parsing `D1`, the method will pick up the `feature:ignore` flag.
* The `D2` class overloads the method that is including via the `using` directive.

What should happen here is a `_wrap_SWIG_0` for the version of the method included via `using`. `wrap_SWIG_n` wrapped versions of the method for any overloads. However, the `feature:ignore` flag set when parsing `D1` was keeping this from happening, so only the overloaded methods were being wrapped, i.e., the method included via the `using` directive was getting missed. This is all explained in more detail with an example in #2872.

This PR fixes this issue by temporarily unsetting the `feature:ignore` flag if and only if that flag was set as a result of `fvirtual`. The way this is tracked is via a new flag called `ignored_by_fvirtual`, which is set whenever `feature:ignore` is set as a result of `fvirtual`. 

There is also a test case included with #2872 that would be nice to include here so that this corner case can be unit tested going forward. I'm not sure of the best way to add this, e.g., the unit test is written for C++ and Python, is there a good way to adapt it so other languages can run the test case as well? Any feedback/pointers on this would be helpful.

wsfulton commented:
Test-suite information: https://swig.org/Doc4.2/Extending.html#Extending_test_suite

I think the reason we haven't found this problem before is because most of the using/method overloading tests have not used virtual methods, so could be a massive category of extra testing that could be added. Just trying to see which testcase to use/enhance...
1. I don't see any director usage in your example, any change with/without directors?
2. Any difference using plain inheritance instead of virtual inheritance?

Some potential testcases to enhance/use:
* using_member_multiple_inherit.i
* using_inherit.i (not more than 2 deep)
* cpp11_using_constructor.i Section: Deeper inheritance chain
* cpp11_director_using_constructor.i

leakec commented:
@wsfulton 
1. In the example there is a `%module(directors="1") test_Py` line near the top of the `test.i` file, so I believe the test is using directors. I also tried the test without directors. Having directors on vs. off does not seem to make a difference. In other words, the test fails with and without directors before this PR and passes with and without directors afterwards.
2. I do not see any change with virtual vs. non-virtual inheritance. In other words, even with non-virtual inheritance the test from #2872 still fails before this PR and still passes afterwards.

Thanks for the link to the test suit info, will read through it. Took a glance and I think enhancing `using_member_multiple_inherit.i` could cover the test from #2872. What I think needs to be added is a `struct` that will not wrap the method under `fvirtual`, i.e., a struct that inherits the method from the `Using1`, but does not do any overloading (call it `Vusing` for now). Then,  another `struct` needs to inherit from `Vusing` (let's just call it `Vusing2`). It needs to have a `using` statement to get the method from `Vusing` and it should overload the method. Then, the Python/Ruby files need to check that both versions of the method are available in `Vusing2`. 

`using_inherit.i` won't work, since as you mentioned it's only 2 deep. 

It's possible this could be an issue with constructors as well. I can check into this when I next have some time (may be a few days). If it is, then enhancing `cpp11_using_constructor.i` and `cpp11_director_using_constructor.i` may also be needed.

leakec commented:
@wsfulton Okay, just added the test case. I added to `using_member_multiple_inherit.i`. It looks like Python and Java are the only ones that use this, so I added the appropriate lines to their `runme` files. I also rolled back `allocate.cxx` to ensure that the text breaks before the changes in this PR. Here were the results:
* Running Python with the old version (before the PR) and without `fvirtual` passes. This is as expected since this was only a problem with `fvritual`.
* Running Python with the old version (before the PR) and with `fvirtual` fails. This is as expected since this is the bug #2872 found. 
* Running Python with the new version (after the PR) and without `fvirtual` passes. This is as expected. Good to check I didn't mess anything up with the new test case.
* Running Python with the new version (after the PR) and with `fvirtual` passes. This is as expected and confirms the PR fixes #2872. 
* Running Java passes in all version (before/after the PR and with/without `fvirtual`). I haven't run much with Java before, but apparently this language doesn't have the same issue that Python did.

leakec commented:
This test change is causing an issue with a specific version of Ruby. Other versions of Ruby are passing. I'm not familiar with Ruby. Is this something anyone can offer advice on?

leakec commented:
> This test change is causing an issue with a specific version of Ruby. Other versions of Ruby are passing. I'm not familiar with Ruby. Is this something anyone can offer advice on?
> 
> Also, the test that is failing is `import_fragments`, which doesn't appear to use any of the code in the test case I changed. However, maybe there is something I'm missing here.

Pinging @wsfulton as I'm pretty unfamiliar with Ruby or why this test might be failing but only on a specific version of Ruby. Do you have anything to recommend other than installing this specific version of Ruby and trying to get a stack trace?

leakec commented:
Just pinging this thread so it doesn't buried. @wsfulton

leakec commented:
Actually, I just realized there is another issue related to ruby testing. According to [this comment](https://github.com/swig/swig/issues/2800#issuecomment-2083391047) Ruby 3.1.5 was having problems, so it was switched to 3.1.4. The checks in this PR ran on April 27th before the change was made to use 3.1.4 on April 29th. Re-running the checks may be enough to fix this.

leakec commented:
Re-running the checks worked! @wsfulton as far as I know, I think everyone has been done in order to merge. Please let me know if there is anything else that is needed.

wsfulton commented:
Hi @leakec. Very nice analysis and the fix looks good to me. Thanks very much, merging shortly.

leakec commented:
@wsfulton Thank you!
